### ISSUE TITLE
Output is different when flan-t5-xl model is converted

### ISSUE THREAD

**Issue by prashantkh19** [OpenNMT/CTranslate2#1487]
I've used the below code to convert and generate the output:

```
model_hf = "google/flan-t5-xl"
model_fast = "ct"

!ct2-transformers-converter --model $model_hf --output_dir $model_fast --force

import ctranslate2
import transformers

fast_model = ctranslate2.Translator(model_fast, device="cuda", device_index=3)
fast_tokenizer = transformers.AutoTokenizer.from_pretrained(model_hf)

num_beams = 1
length_penalty = 1.4
no_repeat_ngram_size = 2
temperature = 0.3
top_k = 100
top_p = 0.92
repetition_penalty = 2.1
max_decoding_length=2048

def get_ctranslate_output(input_text):
    model = fast_model
    tokenizer = fast_tokenizer
    
    def get_input_ids(data):
        return tokenizer(data, max_length=4096, padding="max_length", truncation=True, return_tensors="pt")

    model_inputs = tokenizer.convert_ids_to_tokens(get_input_ids(input_text)["input_ids"][0])
    
    generate_kwargs = dict(
        max_decoding_length=max_decoding_length-1,
        # min_decoding_length=1,
        beam_size=num_beams,
        no_repeat_ngram_size=no_repeat_ngram_size,
        # sampling_topk=top_k,
        # sampling_topp=top_p,
        # sampling_temperature=temperature,
        # length_penalty=length_penalty,
        # repetition_penalty=repetition_penalty
    )
    
    results = model.translate_batch([model_inputs],**generate_kwargs)
    output_tokens = results[0].hypotheses[0]
    output_text = tokenizer.decode(tokenizer.convert_tokens_to_ids(output_tokens))
    return output_text, len(output_tokens)

print(get_ctranslate_output("Generate Essay on Humans: "))

from transformers import T5Tokenizer, T5ForConditionalGeneration, AutoTokenizer, AutoModelForSeq2SeqLM

vanilla_tokenizer = AutoTokenizer.from_pretrained(model_hf)
vanilla_model = AutoModelForSeq2SeqLM.from_pretrained(model_hf,
                                                      device_map="cuda:0", 
                                                      trust_remote_code=True, 
                                                      cache_dir="/home/DATA2/komal/qm/cache")
vanilla_model = vanilla_model.to("cuda")
vanilla_model = vanilla_model.eval()

import torch
def get_vanilla_output(input_text):
    model = vanilla_model
    tokenizer = vanilla_tokenizer
    tokenized_input = tokenizer(input_text,
                          max_length=4096,
                          padding="max_length",
                          truncation=True,
                          return_tensors="pt").to("cuda:0")
    
    with torch.no_grad():
        output = model.generate(**tokenized_input,
                                max_length=max_decoding_length,
                                num_beams = num_beams,
                                no_repeat_ngram_size = no_repeat_ngram_size,
                                # temperature = temperature,
                                # top_k = top_k, 
                                # top_p = top_p
                               )
    result = tokenizer.batch_decode(output, skip_special_tokens=False)
    return result[0], len(output.squeeze())

print(get_vanilla_output("Generate Essay on Humans: "))
```

Output:
```
(':: <unk> - Powered by Blogger.', 12)

('<pad> Humans are the most complex of all animals, and the only one that can be compared to other animals in complexity. Human beings have the capacity to learn, to adapt, change, grow, reproduce, die, live, learn and grow.</s>',
 53)
```

Not sure about what am I doing wrong here.

**Comment by guillaumekln** [comment]
The input of the CTranslate2 model should **not** include padding tokens.

The [T5 example](https://opennmt.net/CTranslate2/guides/transformers.html#t5) shows how to properly tokenize the input text. You just need this line:

```python
model_inputs = tokenizer.convert_ids_to_tokens(tokenizer.encode(input_text))
```

**Comment by prashantkh19** [comment]
Thanks, tried this, and got the same output.

But, when using with `num_beams` > 1, the output changes significantly.

Attaching the response for you reference:
For `num_beams` = 3

```
num_beams = 3
length_penalty = 1.4
no_repeat_ngram_size = 2
temperature = 0.3
top_k = 100
top_p = 0.92
repetition_penalty = 2.1
max_decoding_length=2048

def get_ctranslate_output(input_text):
    model = fast_model
    tokenizer = fast_tokenizer
    
    def get_input_ids(data):
        return tokenizer(data, return_tensors="pt")

    model_inputs = tokenizer.convert_ids_to_tokens(get_input_ids(input_text)["input_ids"][0])
    
    generate_kwargs = dict(
        max_decoding_length=max_decoding_length-1,
        min_decoding_length=1,
        beam_size=num_beams,
        no_repeat_ngram_size=no_repeat_ngram_size,
        # sampling_topk=top_k,
        # sampling_topp=top_p,
        # sampling_temperature=temperature,
        # length_penalty=length_penalty,
        # repetition_penalty=repetition_penalty
    )
    
    results = model.translate_batch([model_inputs],**generate_kwargs)
    output_tokens = results[0].hypotheses[0]
    output_text = tokenizer.decode(tokenizer.convert_tokens_to_ids(output_tokens))
    return output_text, len(output_tokens)

import torch
def get_vanilla_output(input_text):
    model = vanilla_model
    tokenizer = vanilla_tokenizer
    tokenized_input = tokenizer(input_text,
                          # max_length=4096,
                          # padding="max_length",
                          # truncation=True,
                          return_tensors="pt").to("cuda:0")
    
    with torch.no_grad():
        output = model.generate(**tokenized_input,
                                max_length=max_decoding_length,
                                num_beams = num_beams,
                                no_repeat_ngram_size = no_repeat_ngram_size,
                                # temperature = temperature,
                                # top_k = top_k, 
                                # top_p = top_p
                               )
    result = tokenizer.batch_decode(output, skip_special_tokens=True)
    return result[0], len(output.squeeze())

print(get_ctranslate_output("Generate Essay on Humans: "))
print(get_vanilla_output("Generate Essay on Humans: "))
```
Output:

```
('Humans are a species of mammal that evolved from eudicots. They are the closest living relatives of antelopes, deer, and giraffes and the only living members of the genus Hominidae. Hominins have been classified into three subclades: Homo sapiens, Homono habilis (the common ancestor of modern humans), and Homolodonta (a subspecies of Homan).',
 108)

('Humans are a species of mammal that evolved in Africa and Asia. They are the largest land animal and the most diverse animal group on the planet. The human species has been divided into two major groups: monotremes and equidae. Monotreme species are monogamous, mate for life, and reproduce sexually. EQUIDAE species reproduce symbiotically with other species, but do not produce offspring. Phylogenetic analysis of human populations has shown that humans are highly polymorphic, with varying degrees of admixture of different species. This is due to the fact that the human genome is composed of multiple copies of the same alleles.',
 153)
```

**Comment by prashantkh19** [comment]
Update: I'm not getting same output with a complex prompt, even with beam-size=1.

**Comment by guillaumekln** [comment]
For beam search, see https://github.com/OpenNMT/CTranslate2/issues/1493 for an explanation.

In general it is almost impossible to perfectly match the output of other frameworks. Even the same framework can have a different output when run on another system.

Your first error was to include padding in the CTranslate2 input. After fixing, I don't another issue related with CTranslate2.

**Comment by prashantkh19** [comment]
Thanks for the beam search explanation.

However, we debug more on the different output we're getting with greedy search, and we found out that when the input_token length exceeds 1024, the output changes drastically, but if I keep the input_token length less than 1024, the output matches exactly (even at the logits level).

Could you happen to know the issue here?

**Comment by prashantkh19** [comment]
Got it via `max_input_length` parameter. Thanks
